// krakendb/krakendb.js
const fs = require("fs");
const path = require("path");

class db {
    constructor (name, rows, cont) {
        this.name = name;
        this.rows = rows;

        this.data = [['0']];

        if (Array.isArray(rows)) {
            for (let i in rows) {
                this.data[0].push(rows[i]);
            }
        } else {
            throw "The second argument must be an array of strings.";
        }
        if (cont) {
            this.data = cont;
        }
    }
    entry (item) {
        let arr = new Array(this.rows.length + 1);
        arr[0] = item;
        this.data.push(arr);
    }
    set (row, column, val) {
        for(let i in this.data) {
            if(this.data[i][0] == row) {
                if (column) {
                    let colIndex = this.data[0].indexOf(column);
                    if (colIndex == -1)
                        throw "row not found in database"
                    else
                        this.data[i][colIndex] = val;
                        return;
                } else {
                    throw "All 3 arguments are required";
                }
            }
        }
        throw "Something failed";
    }
    get (row, column) {
        for(let i in this.data) {
            if(this.data[i][0] == row) {
                if (column) {
                    let colIndex = this.data[0].indexOf(column);
                    if (colIndex == -1)
                        throw "row not found in database"
                    else
                        return this.data[i][colIndex];
                } else {
                    return this.data[i];
                }
            }
        }
        return false;
    }
    del (item, row) {
        let matches = 0;
        for (let i =0; i < this.data.length; i++) {
            if (this.data[i][0] == item) {

                if (row != undefined) {
                    let matches2 = 0;
                    for (let j = 0; j < this.data[0].length; j++) {
                        if (this.data[0][j] == row) {

                            this.data[i][j] = null;

                            matches2 ++;
                            break;
                        }
                    }
                    if (matches2 == 0) {
                        throw "The selected row is not available"
                    }
                } else {
                    this.data.splice(i, 1);
                }

                matches ++;
                break;
            }
        }
        if (matches == 0) {
            throw "The selected item is not available"
        }
    }
}
var dbs = [];
let activeIndex = 0;
module.exports.newdb = function (name, rows) {
    dbs.push(new db(name, rows));
}
module.exports.activeDB = function(name) {
    let matches = 0;
    for (let i in dbs) {
        if (dbs[i].name == name) {
            activeIndex = i;
            matches++;
        }
    }
    if (matches == 0) {
        throw "The selected database does not exist"
    }
}
module.exports.newEntry = function(name) {
    dbs[activeIndex].entry(name);
}
module.exports.setItem = function(item, row, val) {
    dbs[activeIndex].set(item, row, val);
}
module.exports.getItem = function(item, row) {
    return dbs[0].get(item, row)
}
module.exports.delItem = function(item, row) {
    dbs[activeIndex].del(item, row);
}
module.exports.exportdb = function(db) {
    if (!db) {
        if (fs.existsSync(path.join(__dirname, "krakendb"))) {
            fs.writeFileSync(path.join(path.join(__dirname, "krakendb"), dbs[activeIndex].name + ".json"), JSON.stringify(dbs[activeIndex]));
        } else {
            fs.mkdirSync(path.join(__dirname, 'krakendb'));
            fs.writeFileSync(path.join(path.join(__dirname, "krakendb"), dbs[activeIndex].name + ".json"), JSON.stringify(dbs[activeIndex]));
        }
    } else {
        let matches = 0;
        for (let i in dbs) {
            if (dbs[i].name == db) {
                if (fs.existsSync(path.join(__dirname, "krakendb"))) {
                    fs.writeFileSync(path.join(path.join(__dirname, "krakendb"), dbs[i].name + ".json"), JSON.stringify(dbs[i]));
                } else {
                    fs.mkdirSync(path.join(__dirname, 'krakendb'));
                    fs.writeFileSync(path.join(path.join(__dirname, "krakendb"), dbs[i].name + ".json"), JSON.stringify(dbs[i]));
                }
                matches++;
                break;
            }
        }
        if (matches == 0) {
            throw "The selected database does not exist"
        }
    }
}
module.exports.loaddb = function(name) {
    if (name) {
        if (fs.existsSync(path.join(path.join(__dirname, "krakendb"), name + ".json"))) {

            let rows = JSON.parse(fs.readFileSync(path.join(path.join(__dirname, "krakendb"), name + ".json"))).data[0].slice(1);
            let data = JSON.parse(fs.readFileSync(path.join(path.join(__dirname, "krakendb"), name + ".json"))).data;

            dbs.push(new db(name, rows, data));
            module.exports.activeDB(name);
        } else {
            throw "database doesn't exist";
        }
    } else {
        throw "Please specify a database name (without extension)";
    }
}
module.exports.dbexists = function(name) {
    if (name) {
        if ( fs.existsSync(path.join(path.join(__dirname, "krakendb"), name + ".json")) ) {
            return true;
        } else {
            let matches = 0;
            for (i in dbs.length) {
                if (dbs[i].name == name) {
                    return true;
                }
            }
            if (matches == 0) {
                return false;
            }
        }
    }
}
module.exports.isset = function (row, column) {
    let db = dbs[activeIndex]
    let output = false;
    for(let i in db.data) {
        if(db.data[i][0] == row) {
            if (column) {
                var colIndex = null;
                for(let j in db.data[i]) {
                    if(db.data[0][j] == column) {
                        colIndex = j;
                    }
                }
                if(colIndex != null )
                    output = db.data[i][colIndex] != null;
                else
                    throw "row not found in database"
            } else {
                output = true;
            }
        }
    }
    return output;
}
module.exports.indb = function (str) {
    let db = dbs[activeIndex]
    let matches = 0;
    for (let i in db.data) {
        for (let j in db.data[i]) {
            if (db.data[i][j] == str) {
                matches++
            }
        }
    }
    if (matches == 0)
        return false
    else
        return true;
}
